packages/runtime/datastore/src/localChannelContext.ts

```diff
@@ -119,19 +119,29 @@ export class LocalChannelContext implements IChannelContext {
         this.services.value.deltaConnection.reSubmit(content, localOpMetadata);
     }
 
-    public async snapshot(fullTree: boolean = false): Promise<ITree> {
-        return this.getAttachSnapshot();
-    }
-
-    public async summarize(fullTree: boolean = false): Promise<ISummarizeResult> {
-        const snapshot = this.getAttachSnapshot();
+    /**
+     * Returns a summary at the current sequence number.
+     * @param fullTree - true to bypass optimizations and force a full summary tree
+     * @param trackState - This tells whether we should track state from this summary.
+     */
+    public async summarize(fullTree: boolean = false, trackState: boolean = false): Promise<ISummarizeResult> {
+        assert(this.isLoaded && this.channel !== undefined, "Channel should be loaded to take summary");
```

Comment:
This is probably correct, but worth asking: what ensures that it is already loaded?

Reply:
Ideally, local channel contexts should always be loaded. However, with detached data contexts, this might not be as straightforward as I thought it should be. Looking at the code, it seems there is a possibility that if summarize is called on a detached context, it will hit this assert. But isn't that the right thing to do? Or actually, we should assert way before this in that case.

Reply:
The parent which is datastore should not call this if it is not loaded. So the assert is correct. It should be loaded if we want to summarize it.

Reply:
I don't see any check in the datastore that checks that it is loader before calling summarize. It checks the following:
```
private isChannelAttached(id: string): boolean {
        return (
            // Added in createChannel
            // Removed when bindChannel is called
            !this.notBoundedChannelContextSet.has(id)
            // Added in bindChannel only if this is not attached yet
            // Removed when this is attached by calling attachGraph
            && !this.localChannelContextQueue.has(id)
            // Added in attachChannel called by bindChannel
            // Removed when attach op is broadcast
            && !this.pendingAttach.has(id)
        );
    }
```

Reply:
Those are the conditions whether the channel is attached or not. When we take summary in detached container we use the attach summary path:
            if (!this.notBoundedChannelContextSet.has(objectId)) {
                let snapshot: ITree;
                if (value.isLoaded) {
                    snapshot = value.getAttachSnapshot();
                } else {
                    // If this channel is not yet loaded, then there should be no changes in the snapshot from which
                    // it was created as it is detached container. So just use the previous snapshot.
                    assert(!!this.dataStoreContext.baseSnapshot,
                        "BaseSnapshot should be there as detached container loaded from snapshot");
                    snapshot = convertSnapshotToITree(this.dataStoreContext.baseSnapshot.trees[objectId]);
                }

                // And then store the tree
                entries.push(new TreeTreeEntry(objectId, snapshot));
            }

Reply:
Here we are checking that the channel is local and then checking if it is attached then only call it to summarize.

Reply:
But that's only in getAttachSnapshot(). What happens when summarize() is called on a detached data store context?

Reply:
That is never happens actually. When taking summary in detach container we always use getAttachSummary() path as that is synchronous while summarize is async.